docs/naps/4-async-slicing.md

```diff
@@ -0,0 +1,636 @@
+        - much complexity encapsulated here and other related classes like `ImageSliceData`.
+- `Points`
+    - `__indices_view: Array[int, (M,)]`, indices of points (i.e. rows of `data`) that are in the current slice.
+        - many private properties derived from this (e.g. `_indices_view`, `_view_data`).
+    - `_view_size_scale: Union[float, Array[float, (M,)]]`, used with thick slices of points `_view_size` to make out of slice points appear smaller.
+- `Shapes`
+    - `_data_view: ShapeList`, stores all shapes' data.
+		- `_mesh: Mesh`, stores concatenated meshes of all shapes.
+			- `displayed_triangles: Array[int, (M, 3)]`, triangles to be drawn.
+			- `displayed_triangles_colors: Array[float, (M, 4)]`, per triangle color.
+- `Vectors`
+    - `_view_indices: Array[int, (-1)]`, indices of vectors that are in the current slice.
+        - lots of private properties derived from this (e.g. `_view_face_color`, `_view_faces`).
+
+The vispy layers also read other state from their corresponding layer.
+In particular they read `Layer._transforms` to produce a transform that can be used to properly
+represent the layer in the vispy scene coordinate system.
+
+## Detailed description
+
+The following diagram shows the new proposed approach to slicing layers asynchronously.
+
+![](https://raw.githubusercontent.com/andy-sweet/napari-diagrams/main/napari-slicing-async-calls.drawio.svg)
+
+As with the existing synchronous slicing design, the `Dims.events.current_step` event
+is the shared starting point. In the new approach, we pass `ViewerModel.layers` through
+to the newly defined `LayerSlicer`, which makes a slice request for each layer on the main thread.
+This request is processed asynchronously on a dedicated thread for slicing, while the main thread
+returns quickly to the Qt main event loop, allowing napari to keep responding to other updates
+and interactions. When all the layers have generated slice responses on the slicing thread,
+the `slice_ready` event is emitted. That triggers `QtViewer.on_slice_ready` to be executed on
+the main thread, so that the underlying `QWidgets` can be safely updated.
+
+The rest of this section defines some new types to encapsulate state that is critical
+to slicing and some new methods that redefine the core logic of slicing.
+
+### Request and response
+
+First, we introduce a request to encapsulate the required input to slicing.
+
+```python
+class LayerSliceRequest:
+    data: ArrayLike
+    world_to_data: Transform
+    point: Tuple[float, ...]
+    dims_displayed: Tuple[int, ...]
+    dims_not_displayed: Tuple[int, ...]
+```
+
+At a minimum, this request must contain the data to be sliced, the point at
+which we are slicing in napari's shared world coordinate system, and a way to
+transform world coordinates to a layer's data coordinate system.
+In the future, the point may become a bounding box that includes the current
+field of view.
```

Comment:
I like bbox much better. `point` could just get encoded in the (affine) transform, and the extent information from bbox would be very valuable.

I guess the proposed design is there because that fits the current use, hence "in the future". But imo the future is now :)

Reply:
I would love to collapse a few of the existing request fields into something `bounding_box_world`. The main reason I didn't do that yet is laziness - partly because some of the existing slicing code uses `point` by name and type. But I agree with you that it is probably the right time to make the switch, so I think this would be a great idea when we start writing code to be merged into napari.

Reply:
One things that makes me hesitant is around dimension correspondence. Ideally, the bounding box would have named dimensions and we could do the correspondence between its and a layer's dimensions by name. I think we should still be able to work with the numpy broadcasting correspondence we have now, but I haven't through that yet.